pkg/workloadslicing/workloadslicing.go

```diff
@@ -0,0 +1,314 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package workloadslicing contains definitions to support workload slicing functionality.
+package workloadslicing
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"slices"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/controller/core/indexer"
+	clientutil "sigs.k8s.io/kueue/pkg/util/client"
+	"sigs.k8s.io/kueue/pkg/util/pod"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+// Workload slicing refers to a specialized Kueue feature designed to support workload scaling up.
+// A workload slice represents a logical grouping of two workloads:
+// the "old slice" (preemptible workload) and the "new slice" (preemptor workload).
+
+// Workload slicing can be enabled for a specific integration job instance,
+// provided that the integration supports workload slicing.
+const (
+	// EnabledAnnotationKey refers to the annotation key present on Job's that support
+	// workload slicing.
+	EnabledAnnotationKey = "kueue.x-k8s.io/elastic-job"
+	// EnabledAnnotationValue refers to the annotation value. To enable
+	// workload slicing for a given job, we match both annotation key and value.
+	EnabledAnnotationValue = "true"
+)
+
+// Enabled returns true if a provided job has slicing annotation indicator.
+// Indicator matched on both annotation key and value.
+func Enabled(object metav1.Object) bool {
+	if object == nil {
+		return false
+	}
+	annotations := object.GetAnnotations()
+	return len(annotations) > 0 && annotations[EnabledAnnotationKey] == EnabledAnnotationValue
+}
+
+const (
+	// WorkloadPreemptibleSliceNameKey is the annotation key used to capture an "old" workload slice
+	// that will be preempted by the "new" workload slice, i.e., this annotates a "new" workload slice.
+	WorkloadPreemptibleSliceNameKey = "kueue.x-k8s.io/workload-slice-replacement-for"
+)
+
+// PreemptibleSliceKey returns a preemptible workload slice key if this workload
+// was annotated with such, otherwise, returns an empty string.
+func PreemptibleSliceKey(wl *kueue.Workload) string {
```

Comment:
```suggestion
func PreemptibleSliceKey(wl *kueue.Workload) *workload.Reference {
```
nit, but I would prefer strong typing, it seems easier to use then